src/converters.js

```diff
@@ -984,11 +981,79 @@ function hoistNestedLists( nextIndent, modelRemoveStartPosition, viewRemoveStart
 	// Handle multiple lists. This happens if list item has nested numbered and bulleted lists. Following lists
 	// are inserted after the first list (no need to recalculate insertion position for them).
 	for ( const child of [ ...viewRemovedItem.getChildren() ] ) {
-		if ( child.is( 'ul' ) || child.is( 'ol' ) ) {
+		if ( isList( child ) ) {
 			insertPosition = viewWriter.move( viewWriter.createRangeOn( child ), insertPosition ).end;
 
 			mergeViewLists( viewWriter, child, child.nextSibling );
 			mergeViewLists( viewWriter, child.previousSibling, child );
 		}
 	}
 }
+
+// Checks if view element is a list type (ul or ol).
+//
+// @param {module:engine/view/element~Element} viewElement
+// @returns {Boolean}
+function isList( viewElement ) {
+	return viewElement.is( 'ol' ) || viewElement.is( 'ul' );
+}
+
+// Calculates the indent value for a list item. Handles HTML compliant and non-compliant lists.
+//
+// Also, fixes non HTML compliant lists indents:
+//
+//		before:                                     fixed list:
+//		OL                                          OL
+//		|-> LI (parent LIs: 0)                      |-> LI     (indent: 0)
+//		    |-> OL                                  |-> OL
+//		        |-> OL                                  |
+//		        |   |-> OL                              |
+//		        |       |-> OL                          |
+//		        |           |-> LI (parent LIs: 1)      |-> LI (indent: 1)
+//		        |-> LI (parent LIs: 1)                  |-> LI (indent: 1)
+//
+//		before:                                     fixed list:
+//		OL                                          OL
+//		|-> OL                                      |
+//		    |-> OL                                  |
+//		         |-> OL                             |
+//		             |-> LI (parent LIs: 0)         |-> LI        (indent: 0)
+//
+//		before:                                     fixed list:
+//		OL                                          OL
+//		|-> LI (parent LIs: 0)                      |-> LI         (indent: 0)
+//		|-> OL                                          |-> OL
+//		    |-> LI (parent LIs: 0)                          |-> LI (indent: 1)
+//
+// @param {module:engine/view/element~Element} listItem
+// @param {Object} conversionStore
+// @returns {Number}
+function getIndent( listItem ) {
+	let indent = 0;
+
+	let parent = listItem.parent;
+
+	while ( parent && ( parent.is( 'li' ) || isList( parent ) ) ) {
```

Comment:
Regression:

```
<li>
    text
    <div>
        <ul>
             <li>foo</li>
        </ul>
     </div>
</li>
```

That's the danger of reading the indent from the tree without the context of what we had previously consumed.

If we'll iterate up to the top of the tree, then we may process stuff that got or will get filtered out. If we stop at ul/ol/li, then we ignore some scenarios.

So... I'd unforunately see safer if we relied on stuff that was already converted. In other words, on what we know really is in the model.

Reply:
I tested https://jsfiddle.net/tz8j5Ly6/ on master and it works fine.

Reply:
So we definitively must add this case to the tests then. Looks like I've relied too much on current tests :( I'll look into this and similar problems:
a) no parent ul/ol
b) something wrapping inner ul/ol in the tree

I think that we _could_ calculate this from a tree but differently. But if not doable in a sensible way I'll try to make a solution based on the model (already converter LIs).